### ISSUE TITLE
Windows Client crash after locking screen

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#901]
*Issue migrated from trac ticket # 901*

**component:** client | **priority:** critical | **resolution:** fixed

#### 2015-06-30 00:40:13: maxmylyn created the issue

___
> Versions:
> 
> - Windows 8.1 trunk [r9554](../commit/2c4b40be5a94fb39d72ee22da2cd0445a52b2646) client [http://xpra.org/beta]
> * Fedora 21 [r9656](../commit/a21700d294189a31f254e56c3ec26072b7064177) trunk server [/wiki/Building built from source]
> 
> Server started with:
> 
> - `xpra start :13 --start-new-commands=yes --bind-tcp=0.0.0.0:2200 --start-child=xterm`
> 
> Client started with:
> 
> - `Xpra_cmd.exe attach tcp:ip:2200`
> 
> Server is a Fedora 21 VM
> 
> Client machine is a MacBook Pro running Windows 8.1 ('bootcamped'; not in a VM, I avoid OSX like the plague) with an Intel Iris Pro graphics chipset
> 
> ----
> 
> Repro steps are pretty simple:
> 
> - Connect like normal
> 
> - Lock the screen (WinKey + L)
> 
> - Walk away, allowing the screen to shut off
> 
> - Come back after 10ish minutes 
>  - Unlocking immediately seems to avoid the issue
> 
> - Unlock the computer
> 
> [[br]]
> 
> Result:
> 
> - Xpra has crashed with the following output to the CMD:
> 
> ```
> 
> 2015-06-29 16:11:33,776 unexpected message: WM_NULL / 0 / 0
> 2015-06-29 16:11:36,073 unexpected message: 798 / 0 / 0
> 2015-06-29 16:11:36,589 unexpected message: 798 / 0 / 0
> 2015-06-29 16:11:41,448 do_paint_rgb24 error
> Traceback (most recent call last):
>   File "xpra\client\window_backing_base.pyc", line 276, in do_paint_rgb24
>   File "xpra\client\gl\gl_window_backing_base.pyc", line 563, in _do_paint_rgb24
> 
>   File "xpra\client\gl\gl_window_backing_base.pyc", line 658, in _do_paint_rgb
>   File "xpra\client\gl\gl_window_backing_base.pyc", line 475, in present_fbo
>   File "latebind.pyx", line 44, in OpenGL_accelerate.latebind.Curry.__call__ (c:
> \Users\mcfletch\OpenGL-dev\OpenGL-ctypes\OpenGL_accelerate\src\latebind.c:1201)
>   File "C:\Program Files (x86)\Xpra\OpenGL\GL\exceptional.py", line 41, in glBeg
> in
> WindowsError: exception: access violation writing 0x0B7A5000
> 
> ```
> 
> Server outputs:
> 
> ```
> Disconnecting client '10.0.11.123:63039': client ping timeout (waited 60 seconds without a response)
> xpra client disconnected.
> ```
> 
> And then after a minute or so:
> 
> ```
> 2015-06-29 16:12:36,175 sound-source failed to encode packet: ['info', {'volume': 0, 'state': None, 'codec': '', 'codec_description': u'MPEG-1 Layer 3 (MP3)', 'pid': 10200, 'bytes': 5086536, 'pipeline': 'pulsesrc ! volume name=volume volume=1.0 ! lamemp3enc encoding-engine-quality=0 ! appsink name=sink emit-signals=true max-buffers=10 drop=true sync=false async=false qos=false', 'buffers': 48909}]
> Traceback (most recent call last):
>   File "/usr/lib64/python2.7/site-packages/xpra/net/protocol.py", line 482, in encode
>     main_packet, proto_flags = self._encoder(packet)
>   File "/usr/lib64/python2.7/site-packages/xpra/net/packet_encoding.py", line 64, in do_bencode
>     return bencode(data), FLAGS_BENCODE
>   File "xpra/net/bencode/cython_bencode.pyx", line 204, in xpra.net.bencode.cython_bencode.bencode (xpra/net/bencode/cython_bencode.c:4300)
> ValueError: cannot encode '['info', {'volume': 0, 'state': None, 'codec': '', 'codec_description': u'MPEG-1 Layer 3 (MP3)', 'pid': 10200, 'bytes': 5086536, 'pipeline': 'pulsesrc ! volume name=volume volume=1.0 ! lamemp3enc encoding-engine-quality=0 ! appsink name=sink emit-signals=true max-buffers=10 drop=true sync=false async=false qos=false', 'buffers': 48909}]': found None value!
> ```
> 
> Not sure if the Bencode error is relevant or not.
> 
> In addition, this crash can be induced reliably; even with the same server session. Additional info can be provided upon request. However, I will be out of office from the 1st through the 15th, but I might be able to look at some stuff from home if I get the chance. If there's anything pressing I will keep an eye on my work email periodically(and will have this laptop with me), and if it's super pressing, Alex should be able to test this as well.

**Comment by totaam** [comment]
There were some fixes in the memoryview and opengl areas.
Can you still reproduce?

**Comment by totaam** [comment]
If you can still reproduce it, please run with `-d win32` and try to spot the win32 messages that correspond to going to sleep / activating the screensaver. (and also resuming if possible, you may need to turn off opengl to make it happen without crashing)

We could just minimize all the windows when we detect the event, and restore them later. Unmapped windows don't have an opengl context.

**Comment by totaam** [comment]
You should be able to lower the screen timeout so you don't have to wait 10 minutes to test.

**Comment by totaam** [comment]
There are two events which we should be able to use:
* [WM_DWMNCRENDERINGCHANGED](https://msdn.microsoft.com/en-us/library/windows/desktop/dd388200(v=vs.85).aspx) seems like the right type of message as it is related to rendering, but the [DwmGetWindowAttribute](https://msdn.microsoft.com/en-us/library/windows/desktop/aa969515(v=vs.85).aspx) call is too obscure to use.
* [WM_WTSSESSION_CHANGE](https://msdn.microsoft.com/en-us/library/aa383828(v=vs.85).aspx): *Notifies applications of changes in session state.*. From your log, it seems to go into `WTS_SESSION_LOCK` then 3 seconds later `WTS_SESSION_UNLOCK`. It is less clean than the event above because we have to infer that the opengl rendering context is going to be broken by the lock + unlock. (which may not be true for all OS versions and drivers... but that's what we have available)

**Comment by totaam** [comment]
Some preparatory / cleanup work in [r10760](../commit/71fabd075e60289e0f21599b8150ecae2edd8beb) + a fix in [r10761](../commit/5b1d4cf69ecd61eef9622130dadbd2de1cf5414f). (may backport those)

[r10762](../commit/3751f634103aa2b3efa27d21d4643299658968fe) should fix this: we now iconify our windows when the session is locked on win32, so the windows are "umapped" and we close the OpenGL context.
You will then have to unminimize them when you unlock.

Note: assuming that this does fix things for you (I understand it may take time to confirm as it is harder to confirm a negative result... side note: I believe a fast updating window like glxgears is more likely to trigger this problem), then we can:
* try to unminimize the windows when we get the corresponding unlock event (will have to add a new state to keep track of which windows need to be unminimized)
* prevent new windows from popping up whilst the screen is locked, which is not implemented yet as this is harder and also less likely to occur (we could fake the configure event and send a "minimize" message immediately after, but this would not take care of various other messages which can cause the minimized window to be mapped... and therefore may be able to trigger the crash)

I am raising the priority of this ticket because I would like to get some kind of fix in 0.15

**Comment by totaam** [comment]
Ah, but of course: iconifying the window is not enough, we still have a widget that references the OpenGL context and we still paint to it even, albeit at a slower rate (so that the window manager can show previews in the task bar and show the window contents as soon as we we deiconify it)

So [r10778](../commit/6e26eece2b3bb6f2eabc0c2886864a3709a3d829) + [r10780](../commit/02d7d60ad59dbb80236fd926cc5bf75445242e95) introduces a "freeze" call on the window class which we use to free the OpenGL context by removing the widget then re-creating one next time we need it again. This is all a bit too involved to backport to v0.15.x.

To make it easier to test this code, especially on other platforms where locking does not cause the windows to be frozen, you can still trigger it with:
```
--key-shortcut=Meta+Shift+F7:freeze
```

[r10779](../commit/190a346be50d659e4ec7891a8da6bdb90a922052) should make it very rare (though maybe not impossible? but no big deal) to get paint errors reported as we finish to decompress a picture after we've destroyed the window's opengl backing context: those should now get logged at debug level only.


Things I would like to do more properly (which I will need to put in a new ticket):
* as mentioned in comment:8 there is nothing stopping applications from creating new windows (or showing / resizing existing ones) after we call freeze, which may still cause the crash...
* maybe the "_been_mapped" flag should be reset to False when we freeze (or unmap?) a window
* restoring the windows when the screen is unlocked


@maxmylyn: does this work for you?

**Comment by totaam** [comment]
I have moved the unresolved issues to #997. (restore windows, block new windows)

Found one remaining issue `AssertionError` thanks to your client log output, it should be fixed by [r10786](../commit/0a91bd9b65266c567e71e600180662098c0752d1). This could have caused this minimize code to fail for some windows in some cases! ..and the crash to re-appear.

Closing at last!

(I am not backporting these changes for now: just too many changes, too intrusive, let's focus on 0.16)
